packages/openapi-v3/src/controller-spec.ts

```diff
@@ -214,47 +209,84 @@ function resolveControllerSpec(constructor: Function): ControllerSpec {
 
     for (const p of paramTypes) {
       if (isComplexType(p)) {
-        if (!spec.components) {
-          spec.components = {};
-        }
-        if (!spec.components.schemas) {
-          spec.components.schemas = {};
-        }
-        if (p.name in spec.components.schemas) {
-          // Preserve user-provided definitions
-          debug(
-            '    skipping parameter type %j as already defined',
-            p.name || p,
-          );
-          continue;
-        }
-        const jsonSchema = getJsonSchema(p);
-        const openapiSchema = jsonToSchemaObject(jsonSchema);
-        const outputSchemas = spec.components.schemas;
-        if (openapiSchema.definitions) {
-          for (const key in openapiSchema.definitions) {
-            // Preserve user-provided definitions
-            if (key in outputSchemas) continue;
-            const relatedSchema = openapiSchema.definitions[key];
-            debug(
-              '    defining referenced schema for %j: %j',
-              key,
-              relatedSchema,
-            );
-            outputSchemas[key] = relatedSchema;
-          }
-          delete openapiSchema.definitions;
-        }
-
-        debug('    defining schema for %j: %j', p.name, openapiSchema);
-        outputSchemas[p.name] = openapiSchema;
-        break;
+        generateOpenAPISchema(spec, p);
       }
     }
   }
   return spec;
 }
 
+/**
+ * Resolve the x-ts-type in schema objects
+ * @param spec Controller spec
+ * @param content Parent object of the schema
+ */
+function resolveTSType(
+  spec: ControllerSpec,
+  content: {[TS_TYPE_KEY]?: Function; schema?: SchemaObject},
+) {
+  debug('  evaluating content: %j', content);
+  const tsType = content[TS_TYPE_KEY];
+  debug('  %s => %o', TS_TYPE_KEY, tsType);
+  if (tsType) {
+    content.schema = resolveSchema(tsType, content.schema);
+    if (content.schema.$ref) generateOpenAPISchema(spec, tsType);
+
+    // We don't want a Function type in the final spec.
+    delete content[TS_TYPE_KEY];
+  } else {
+    const schema = content.schema;
+    if (!schema || isReferenceObject(schema)) return;
+    if (schema.type === 'array') {
+      const itemSchema = schema.items;
+      if (!itemSchema || isReferenceObject(itemSchema)) return;
+      const itemTsType = itemSchema[TS_TYPE_KEY];
+      debug('  %s => %o', TS_TYPE_KEY, itemTsType);
+      if (itemTsType) {
+        schema.items = resolveSchema(itemTsType, itemSchema);
+        if (schema.items.$ref) generateOpenAPISchema(spec, itemTsType);
+        // We don't want a Function type in the final spec.
+        delete schema.items[TS_TYPE_KEY];
```

Comment:
AFAICT, these three lines (or even more than three) are copy&pasted from the case when schema is not an array. Can we share this code please?

Also is it possible for an array to contain another nested array? E.g. an array of arrays of x-ts-type schemas.

```yaml
schema:
 type: array
 items: 
   type: array
   items:
    x-ts-type: Number
```

I think it would be best if we could recursively call `resolveTSType` to handle items.

```ts
resolveTSType(spec, schema.items);
```